dockerversion/useragent.go

```diff
@@ -20,5 +32,24 @@ func DockerUserAgent() string {
 	httpVersion = append(httpVersion, useragent.VersionInfo{Name: "os", Version: runtime.GOOS})
 	httpVersion = append(httpVersion, useragent.VersionInfo{Name: "arch", Version: runtime.GOARCH})
 
-	return useragent.AppendVersions("", httpVersion...)
+	return upstreamUA + useragent.AppendVersions("", httpVersion...)
+}
+
+// SaveUserAgentContext adds the user-agent string ua to the context ctx
+func SaveUserAgentContext(ctx *context.Context, ua string) {
+	var k uaStringKey
+	*ctx = context.WithValue(*ctx, k, ua)
+}
```

Comment:
nit: For consistency with other context-related functions, call this `WithUserAgent` (or `ContextWithUserAgent`?) and have it return the new context instead of assigning to a pointer argument.